File: LC/Python/Medium/1801.py
class Solution:
    def getNumberOfBacklogOrders(self, orders: List[List[int]]) -> int:
        buy_orders = []
        sell_orders = []
        
        for price, amount, t in orders:
            if t == 0:
                heappush(buy_orders, [-price, amount])
            else:
                heappush(sell_orders, [price, amount])
            
            while buy_orders and sell_orders and sell_orders[0][0] <= -buy_orders[0][0]:
                quantity = min(buy_orders[0][1], sell_orders[0][1])
                buy_orders[0][1] -= quantity
                sell_orders[0][1] -= quantity
                if buy_orders[0][1] == 0:
                    heappop(buy_orders)
                if sell_orders[0][1] == 0:
                    heappop(sell_orders)
                    
        return sum(amount for price, amount in buy_orders + sell_orders) % (10**9 + 7)

File: LC/C++/Medium/1801.cpp
class Solution {
public:
    int getNumberOfBacklogOrders(vector<vector<int>>& orders) {
        std::priority_queue<std::pair<int, int>> buys{};
        std::priority_queue<std::pair<int,int>, std::vector<pair<int,int>>, std::greater<std::pair<int,int>>> sells;
        for (auto& order : orders) {
            int price = order[0], amount = order[1], t = order[2];
            if (t == 0) {
                buys.push({price, amount});
            } else {
                sells.push({price, amount});
            } 

            while (!buys.empty() && !sells.empty() && sells.top().first <= buys.top().first) {
                auto [bp, ba] = buys.top();  buys.pop();
                auto [sp, sa] = sells.top(); sells.pop();

                int quantity = min(ba, sa);
                
                ba -= quantity;
                sa -= quantity;
                
                if (ba > 0) buys.push({bp, ba});
                if (sa > 0) sells.push({sp, sa});
            }
        }

        int result = 0;
        const int MOD = 1e9 + 7;
        
        while (!buys.empty()) {
            result = (result + buys.top().second) % MOD;
            buys.pop();
        }
        while (!sells.empty()) {
            result = (result + sells.top().second) % MOD;
            sells.pop();
        }
        return result;
    }
};